lib/spack/spack/solver/asp.py

```diff
@@ -751,6 +751,8 @@ def key_fn(version):
         most_to_least_preferred = sorted(declared_versions, key=key_fn)
 
         for weight, declared_version in enumerate(most_to_least_preferred):
+            if pkg.name == 'yaml-cpp':
+                print(weight, declared_version.version, declared_version.origin)
```

Comment:
@alalazo @becker33 or @tgamblin  I'm trying understand what is going on with the concretizer. I've been running the following test:
```console
spack solve yaml-cpp@develop=f3c7206350ac8ee364af687deaae5c574dcfca2c
0 develop=f3c7206350ac8ee364af687deaae5c574dcfca2c 0
1 0.7.0 3
2 0.6.3 3
3 0.6.2 3
4 0.5.3 3
5 0.3.0 3
6 develop 3
```
Which seems to indicate that my supplied version will be the most preferred and should be selected by the concretizer. However the results are the opposite:
```console
==> Best of 6 considered solutions.
==> Optimization Criteria:
  Priority  Criterion                                            Installed  ToBuild
  1         number of input specs not concretized                        -        0
  2         number of packages to build (vs. reuse)                      -        1
  3         deprecated versions used                                     0        0
  4         version weight                                               0        6
  5         number of non-default variants (roots)                       0        0
  6         preferred providers for roots                                0        0
  7         default values of variants not being used (roots)            0        0
  8         number of non-default variants (non-roots)                   0        0
  9         preferred providers (non-roots)                              0        0
  10        compiler mismatches                                          0        0
  11        OS mismatches                                                0        0
  12        non-preferred OS's                                           0        0
  13        version badness                                             11        6
  14        default values of variants not being used (non-roots)        0        0
  15        non-preferred compilers                                      0        0
  16        target mismatches                                            0        0
  17        non-preferred targets                                       30       30

yaml-cpp@develop%apple-clang@13.0.0~ipo+pic+shared~tests build_type=RelWithDebInfo arch=darwin-bigsur-skylake
    ^cmake@3.21.3%apple-clang@13.0.0~doc+ncurses+ownlibs~qt build_type=Release arch=darwin-bigsur-skylake
```
so the least preferred version is getting selected?  I'm trying to understand the asp output but it is slow going. Would one of you be able to help me understand what is going on?

Reply:
Ah I think I just figured it out. The problem was the version was considered equivalent with `develop` if I change. Not sure I understand why the less preferred version was getting selected due to the equivalence, but if I make it so the `Version.version=Version.commit_hash` then the concretizer picks my new version syntax.